rules/mapping.go

```diff
@@ -118,7 +165,94 @@ func newMappingRule(
 	}, nil
 }
 
-// ActiveSnapshot returns the latest snapshot whose cutover time is earlier than or
+func newMappingRuleFromFields(
+	name string,
+	rawFilters map[string]string,
+	policies []policy.Policy,
+	cutoverTime int64,
+	opts filters.TagsFilterOptions,
+) (*mappingRule, error) {
+	mr := mappingRule{uuid: uuid.New()}
+	mr.addSnapshot(name, rawFilters, policies, cutoverTime, opts)
+	return &mr, nil
+}
+
+func (mc *mappingRule) Name() (string, error) {
+	if len(mc.snapshots) == 0 {
+		return "", errNoSnapshots
+	}
+	latest := mc.snapshots[len(mc.snapshots)-1]
+	return latest.name, nil
+}
+
+func (mc *mappingRule) Tombstoned() bool {
+	if len(mc.snapshots) == 0 {
+		return true
+	}
+	latest := mc.snapshots[len(mc.snapshots)-1]
+	return latest.tombstoned
+}
+
+func (mc *mappingRule) addSnapshot(
+	name string,
+	rawFilters map[string]string,
+	policies []policy.Policy,
+	cutoverTime int64,
+	opts filters.TagsFilterOptions,
+) error {
+	filter, err := filters.NewTagsFilter(rawFilters, filters.Conjunction, opts)
+	if err != nil {
+		return err
+	}
+	snapshot := &mappingRuleSnapshot{
+		name:         name,
+		tombstoned:   false,
+		cutoverNanos: cutoverTime,
+		filter:       filter,
+		policies:     policies,
+		rawFilters:   rawFilters,
+	}
+
+	mc.snapshots = append(mc.snapshots, snapshot)
+	return nil
+}
+
+func (mc *mappingRule) markTombstoned(cutoverTime int64) error {
+	n, err := mc.Name()
+	if err != nil {
+		return err
+	}
+
+	if mc.Tombstoned() {
+		return fmt.Errorf("%s is already tombstoned", n)
+	}
+
+	snapshot := *mc.snapshots[len(mc.snapshots)-1]
```

Comment:
The call to mc.Name() will fail if there are no snapshots, I didn't want to do this same check twice.